radicle-node/src/wire/new.rs

```diff
@@ -0,0 +1,273 @@
+                );
+            }
+        }
+    }
+
+    fn handle_command(&mut self, cmd: Self::Command) {
+        self.inner.command(cmd)
+    }
+
+    fn handle_error(
+        &mut self,
+        err: reactor::Error<net::SocketAddr, XkAddr<NodeId, net::SocketAddr>>,
+    ) {
+        match err {
+            reactor::Error::ListenerUnknown(id) => {
+                log::error!("asking for unknown listener {}", id);
+            }
+            reactor::Error::PeerUnknown(id) => {
+                log::error!("asking for unknown peer {}", id);
+            }
+            reactor::Error::PeerDisconnected(addr, err) => {
+                log::error!("the peer {} got disconnected with {}", addr, err);
+                self.inner_queue
+                    .push_back(reactor::Action::UnregisterTransport(addr))
+            }
+        }
+    }
+}
+
+impl<R, S, W, G> Iterator for Wire<R, S, W, G>
+where
+    R: routing::Store,
+    S: address::Store,
+    W: WriteStorage + 'static,
+    G: Signer,
+{
+    type Item = reactor::Action<NetAccept<Session>, NetTransport<Session, Framer>>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some(event) = self.inner_queue.pop_front() {
+            return Some(event);
+        }
+
+        match self.inner.next() {
+            Some(Io::Write(id, msgs)) => {
+                log::debug!("Sending {} messages to {}", msgs.len(), id);
+                Some(reactor::Action::Send(id, msgs))
+            }
+            Some(Io::Event(e)) => {
+                log::warn!(
+                    "Received an event, while events are not handled in the current version"
+                );
+                // TODO: Handle the events
+                None
+            }
+            Some(Io::Connect(id, addr)) => {
+                let NetAddr { host, port } = *addr;
+                let socket_addr = match host {
+                    HostAddr::Ip(ip) => net::SocketAddr::new(ip, addr.port()),
+                    HostAddr::Dns(_) => todo!(),
+                    _ => self.proxy,
+                };
+
+                self.inner.attempted(&socket_addr);
+
+                match NetTransport::<Session, Framer>::connect(
+                    PeerAddr::new(id, socket_addr),
+                    &self.local_node,
+                ) {
+                    Ok(transport) => {
+                        self.inner
+                            .connecting(socket_addr, &transport.local_addr(), Link::Outbound);
+                        Some(reactor::Action::RegisterTransport(transport))
+                    }
+                    Err(err) => {
+                        self.inner
+                            .disconnected(&id, &DisconnectReason::DialError(Arc::new(err)));
+                        return None;
+                    }
+                }
+            }
+            Some(Io::Disconnect(id, r)) => {
+                self.inner.disconnected(&id, &DisconnectReason::Protocol(r));
```

Comment:
Same here we should only call this when something actually happened, and not in the same function that is reading from the inner service.